jina/drivers/reduce.py

```diff
@@ -23,73 +22,65 @@ class ReduceAllDriver(RecursiveMixin, BaseRecursiveDriver):
+                    else:
+                        doc_pointers[context_id].matches.extend(docs)
+            else:
+                self._apply_root(docs)
 
 
-class CollectEvaluationDriver(ReduceAllDriver):
+class CollectEvaluationDriver(FlatRecursiveMixin, BaseRecursiveDriver):
     """Merge all evaluations into one, grouped by ``doc.id`` """
 
+    def __init__(self, traversal_paths: Tuple[str] = ('r',), *args, **kwargs):
+        super().__init__(traversal_paths=traversal_paths, *args, **kwargs)
+
     def _apply_all(
             self,
             docs: 'DocumentSet',
-            context_doc: 'Document',
-            field: str,
             *args,
             **kwargs) -> None:
-        if context_doc.id not in self.doc_pointers:
-            self.doc_pointers[context_doc.id] = context_doc.evaluations
-        else:
-            self.doc_pointers[context_doc.id].extend(context_doc.evaluations)
+        doc_pointers = {}
+        for doc in docs:
+            if doc.id not in doc_pointers:
+                doc_pointers[doc.id] = doc.evaluations
+            else:
+                doc_pointers[doc.id].extend(doc.evaluations)
+
 
+class ConcatEmbedDriver(BaseRecursiveDriver):
+    """Concat all embeddings into one, grouped by ``doc.id`` """
 
-class ConcatEmbedDriver(ReduceAllDriver):
-    """Concat all embeddings into one, grouped by ```doc.id``` """
+    def __init__(self, traversal_paths: Tuple[str] = ('r',), *args, **kwargs):
+        super().__init__(traversal_paths=traversal_paths, *args, **kwargs)
 
     def __call__(self, *args, **kwargs):
-        # all pointers of the docs, provide the weak ref to all docs in partial reqs
-        self.doc_pointers = {}  # type: Dict[str, Any]
-        self._traverse_apply(self.docs, *args, **kwargs)
-        self._traverse_apply(self.req.docs, concatenate=True, *args, **kwargs)
-        self.doc_pointers.clear()
+        all_documents = self.docs.traverse_flatten(self._traversal_paths)
+        doc_pointers = self._collect_embeddings(all_documents)
 
-    def _apply_all(
-            self,
-            docs: 'DocumentSet',
-            context_doc: 'Document',
-            field: str,
-            concatenate: bool = False,
-            *args,
-            **kwargs):
-        doc = context_doc
-        if concatenate:
-            doc.embedding = np.concatenate(self.doc_pointers[doc.id], axis=0)
-        else:
-            if doc.id not in self.doc_pointers:
-                self.doc_pointers[doc.id] = [doc.embedding]
-            else:
-                self.doc_pointers[doc.id].append(doc.embedding)
+        last_request_documents = self.req.docs.traverse_flatten(self._traversal_paths)
+        self._concat_apply(last_request_documents, doc_pointers)
+
+    def _collect_embeddings(self, docs: 'DocumentSet'):
+        doc_pointers = defaultdict(list)
+        for doc in docs:
+            doc_pointers[doc.id].append(doc.embedding)
```

Comment:
Where do we know the embeddings are appended in the right order?

Reply:
Ouch, fair point. But this problem is not introduced with the changes of this PR. I create an issue.

Reply:
Ah no, we rely, that the chunks are in the right order for all documents.